File: backend/laravel/laravel_react/app/Services/RatingService.php
<?php

namespace App\Services;

use App\Models\Execution;
use App\Models\Rating;

class RatingService
{
    public function listForUser(int $userId, ?int $executionId): array
    {
        $q = Rating::query()->where('user_id', $userId);
        if ($executionId !== null) {
            $q->where('execution_id', $executionId);
        }

        $rows = $q->orderByDesc('created_at')->get();

        return [
            'items' => $rows->map(fn (Rating $r) => $this->toArray($r))->all(),
        ];
    }

    public function create(int $userId, array $data): array
    {
        $ex = Execution::where('user_id', $userId)->whereKey($data['execution_id'])->first();
        if (! $ex) {
            throw new \InvalidArgumentException('Execution not found');
        }

        if (Rating::where('user_id', $userId)->where('execution_id', $data['execution_id'])->exists()) {
            throw new \InvalidArgumentException('Already rated this execution');
        }

        $r = Rating::create([
            'user_id' => $userId,
            'execution_id' => $data['execution_id'],
            'score' => $data['score'],
            'comment' => $data['comment'] ?? null,
            'created_at' => now(),
        ]);

        return $this->toArray($r);
    }

    private function toArray(Rating $r): array
    {
        return [
            'id' => $r->id,
            'user_id' => $r->user_id,
            'execution_id' => $r->execution_id,
            'score' => $r->score,
            'comment' => $r->comment,
            'created_at' => $r->created_at?->toIso8601String(),
        ];
    }
}

File: backend/flask/flask_react/app/services/rating_service.py
from __future__ import annotations

from app.extensions import db
from app.models import Execution, Rating


def list_ratings(user_id: int, execution_id: int | None = None) -> dict:
    q = Rating.query.filter_by(user_id=user_id)
    if execution_id is not None:
        q = q.filter_by(execution_id=execution_id)
    rows = q.order_by(Rating.created_at.desc()).all()
    return {"items": [_rating_dict(r) for r in rows]}


def create_rating(user_id: int, data: dict) -> dict:
    ex = Execution.query.filter_by(id=data["execution_id"], user_id=user_id).first()
    if not ex:
        raise ValueError("Execution not found")
    existing = Rating.query.filter_by(
        user_id=user_id, execution_id=data["execution_id"]
    ).first()
    if existing:
        raise ValueError("Already rated this execution")
    r = Rating(
        user_id=user_id,
        execution_id=data["execution_id"],
        score=data["score"],
        comment=data.get("comment"),
    )
    db.session.add(r)
    db.session.commit()
    return _rating_dict(r)


def _rating_dict(r: Rating) -> dict:
    return {
        "id": r.id,
        "user_id": r.user_id,
        "execution_id": r.execution_id,
        "score": r.score,
        "comment": r.comment,
        "created_at": r.created_at.isoformat() + "Z",
    }
